pkg/kubelet/network/dns/dns.go

```diff
@@ -40,7 +40,7 @@ import (
 
 var (
 	// The default dns opt strings.
-	defaultDNSOptions = []string{"ndots:5"}
+	defaultDNSOptions = []string{"ndots:5", "single-request"}
```

Comment:
This does not seem a strong evidence to change a default, specially if this only happens with IPVS

An alternative can be to install a webhook that sets this options in all your pods https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config

Reply:
Yes, It's a solution using webhook.

I want to take more discussion.
After check the doc here: https://kubernetes.io/docs/reference/config-api/kube-proxy-config.v1alpha1/#kubeproxy-config-k8s-io-v1alpha1-ProxyMode
Kubeproxy only support 'iptables' and 'ipvs' on linux.

And I found this issue is also there on my very old cluster with k8s version: 1.11.4 . The old cluster kubeproxy is driven by iptables.

Which means this issue is on both (all) currently supported kube-proxy mode. So I think a k8s native solution for this issue is a good idea. Instead of solving it by another webhook. It can make new comer easy to use k8s with less issue.

Because I think it's a k8s native issue. Although it's cause by compatibility issues with glibc and iptables/kernel. It's a good idea to using a workaround by default to avoid this issue.

How do you think about that @aojea

Reply:
it seems this is not supported in muslc https://wiki.musl-libc.org/functional-differences-from-glibc.html so we can not default something that will not work for all the workloads

> If musl’s resolver is requested to translate a hostname to IPv4 and IPv6, both requests will use the same socket, same as glibc. glibc could be configured to send those requests sequentially, with single-request and single-request-reopen options, but musl cannot.

I still think we need more evidence, AFAIK the kernel has fixed all these issues with the racy conntrack entries, and if not we should try to fix it there first and not workaround it

Reply:
En, it's so sad. After doing some research, It looks like Alpine use musl. And musl don't support `single-request` option.
And another sad thing is when I try to force it use TCP by option `use-vc`. musl don't support it either.

And you say that `the kernel has fixed all these issues`. Do you remember which kernel version fixed it?

When I searching. I found here: https://github.com/0xERR0R/blocky/issues/1242 says that Alpine 3.18 which use Linux kernel 6.1 still have this issue.

BTW: If the kernel fixed the issue. Do we need to upgrade host machine kernel or guest docker image kernel?

Reply:
> Do we need to upgrade host machine kernel 

docker containers use the host kernel , the patches are in this doc https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts